crates/red_knot_module_resolver/src/path.rs

```diff
@@ -1,676 +1,286 @@
-    }
-
-    /// Returns `None` if the path doesn't exist, isn't accessible, or if the path points to a directory.
-    pub(crate) fn to_file(&self, search_path: &Self, resolver: &ResolverState) -> Option<File> {
-        ModulePathRef::from(self).to_file(search_path, resolver)
-    }
-
-    pub(crate) fn as_system_path(&self) -> Option<&SystemPathBuf> {
-        match &self.0 {
-            ModulePathBufInner::Extra(path) => Some(path),
-            ModulePathBufInner::FirstParty(path) => Some(path),
-            ModulePathBufInner::StandardLibrary(_) => None,
-            ModulePathBufInner::SitePackages(path) => Some(path),
-            ModulePathBufInner::EditableInstall(path) => Some(path),
-        }
-    }
+pub(crate) struct ModulePath {
+    search_path: SearchPath,
+    relative_path: Utf8PathBuf,
 }
 
-impl fmt::Debug for ModulePathBuf {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match &self.0 {
-            ModulePathBufInner::Extra(path) => {
-                f.debug_tuple("ModulePathBuf::Extra").field(path).finish()
-            }
-            ModulePathBufInner::FirstParty(path) => f
-                .debug_tuple("ModulePathBuf::FirstParty")
-                .field(path)
-                .finish(),
-            ModulePathBufInner::SitePackages(path) => f
-                .debug_tuple("ModulePathBuf::SitePackages")
-                .field(path)
-                .finish(),
-            ModulePathBufInner::StandardLibrary(path) => f
-                .debug_tuple("ModulePathBuf::StandardLibrary")
-                .field(path)
-                .finish(),
-            ModulePathBufInner::EditableInstall(path) => f
-                .debug_tuple("ModulePathBuf::EditableInstall")
-                .field(path)
-                .finish(),
-        }
+#[must_use]
+fn stdlib_path_to_module_name(relative_path: &Utf8Path) -> Option<ModuleName> {
+    let parent_components = relative_path
+        .parent()?
+        .components()
+        .map(|component| component.as_str());
+    let skip_final_part = relative_path.ends_with("__init__.pyi");
+    if skip_final_part {
+        ModuleName::from_components(parent_components)
+    } else {
+        ModuleName::from_components(parent_components.chain(relative_path.file_stem()))
     }
 }
 
-impl PartialEq<SystemPathBuf> for ModulePathBuf {
-    fn eq(&self, other: &SystemPathBuf) -> bool {
-        ModulePathRef::from(self) == **other
-    }
+#[must_use]
+fn query_stdlib_version(
+    custom_stdlib_root: Option<&SystemPath>,
+    relative_path: &Utf8Path,
+    resolver: &ResolverState,
+) -> TypeshedVersionsQueryResult {
+    let Some(module_name) = stdlib_path_to_module_name(relative_path) else {
+        return TypeshedVersionsQueryResult::DoesNotExist;
+    };
+    let ResolverState {
+        db,
+        typeshed_versions,
+        target_version,
+    } = resolver;
+    typeshed_versions.query_module(*db, &module_name, custom_stdlib_root, *target_version)
 }
 
```

Comment:
Nit: This methods between the struct declaration and the `impl` block are a bit distracting to the reading flow. Maybe move them further down?